client/lib/wp/sync-handler/test/cache-index.js

```diff
@@ -0,0 +1,166 @@
+				} );
+			},
+			removeItem( key ) {
+				return new Promise( resolve => {
+					delete localData[ key ];
+					resolve();
+				} );
+			},
+			keys() {
+				return new Promise( resolve => {
+					resolve( Object.keys( localData ) );
+				} )
+			}
+		}
+	}
+};
+
+describe( 'cache-index', () => {
+	before( () => {
+		mockery.enable( {
+			warnOnReplace: false,
+			warnOnUnregistered: false
+		} );
+		mockery.registerMock( 'lib/localforage', localforageMock );
+		cacheIndexModule = rewire( '../cache-index' );
+		( { generateKey } = require( '../' ) );
+		cacheIndex = cacheIndexModule.cacheIndex;
+		RECORDS_LIST_KEY = cacheIndexModule.__get__( 'RECORDS_LIST_KEY' );
+	} );
+
+	after( function() {
+		mockery.disable();
+	} );
+
+	beforeEach( clearLocal ); // also do inside nested blocks with >1 test
+
+	describe( '#getAll', () => {
+		beforeEach( clearLocal );
+
+		it( 'should return empty array for empty localforage', () => {
+			return cacheIndex.getAll()
+			.then( ( res ) => {
+				expect( res ).to.equal( undefined );
+			} );
+		} );
+		it( 'should return index from localforage and nothing else', () => {
+			localData = {
+				test: 1,
+				test2: 2,
+			};
+			localData[ RECORDS_LIST_KEY ] = 3;
+			return cacheIndex.getAll()
+			.then( ( res ) => {
+				expect( res ).to.equal( 3 );
+			} );
+		} );
+	} );
+
+	describe( '#addItem', () => {
+		it( 'should add item to empty index', () => {
+			return cacheIndex.addItem( 'testKey' )
+			.then( () => {
+				const currentIndex = localData[ 'records-list' ];
+				expect( currentIndex ).to.be.an( 'array' );
+				expect( currentIndex[0] ).to.have.property( 'key', 'testKey' );
+				expect( currentIndex[0] ).to.have.property( 'timestamp' );
+			} );
+		} );
+	} );
+
+	describe( '#removeItem', () => {
+		it( 'should remove item from a populated index', () => {
+			setRecordsList( [
+				{ key: 'key1' },
+				{ key: 'key2' },
+				{ key: 'key3' },
+			] );
+			return cacheIndex.removeItem( 'key2' )
+			.then( () => {
+				const currentIndex = localData[ 'records-list' ];
+				expect( currentIndex ).to.eql( [ { key: 'key1' }, { key: 'key3' } ] );
+			} );
+		} );
+	} );
+
+	describe( '#pruneRecordsFrom', () => {
+		it( 'should prune old records', () => {
+			const now = Date.now();
+			const yesterday = now - ms( '1 day' );
+			localData = {
+				key1: 'test1',
+				key2: 'test2',
+			};
+			setRecordsList( [
+				{ key: 'key1', timestamp: now },
+				{ key: 'key2', timestamp: yesterday },
+			] );
+			return cacheIndex.pruneRecordsFrom( '1 hour' )
+			.then( () => {
+				const currentIndex = localData[ 'records-list' ];
+				expect( currentIndex ).to.eql( [ { key: 'key1', timestamp: now } ] );
+				expect( localData ).to.have.property( 'key1', 'test1' );
+				expect( localData ).to.have.property( RECORDS_LIST_KEY );
+				expect( localData ).to.not.have.property( 'key2' );
+			} );
+		} );
+	} );
+
+	describe( '#clearAll', () => {
+		it( 'should clear all sync records and nothing else', () => {
```

Comment:
I think we may be ok in this situation since the mocks resolve immediately, but if you need to test real async stuff, you can pass through done:
```javascript
it( 'should clear all sync records and nothing else', ( done ) => {
```
And call ```done()```it when you're finished in the callback. If that gets too slow, then async stuff should be moved to test setup, and the tests themselves check the result value.